Support Rosetta with VZ driver

chancez opened the pull request:
Useful reading: https://developer.apple.com/documentation/virtualization/running_intel_binaries_in_linux_vms_with_rosetta

Implementation based off the sample here: https://github.com/Code-Hex/vz/tree/9de1c1041d5e560f46f9f94fc499a1cc72fd92e5/example/gui-linux

AkihiroSuda commented on examples/experimental/vz.yaml:
```diff
@@ -11,3 +16,13 @@ mounts:
+    /usr/sbin/update-binfmts --install rosetta /mnt/lima-rosetta/rosetta \
+    --magic "\x7fELF\x02\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x3e\x00" \
+    --mask "\xff\xff\xff\xff\xff\xfe\xfe\x00\xff\xff\xff\xff\xff\xff\xff\xff\xfe\xff\xff\xff" \
+    --credentials yes --preserve no --fix-binary yes
```
- No need to install binfmt-support if you write to `/proc/sys/fs/binfmt_misc` by yourself
- I'd expect this to be moved to a YAML property like `rosetta.binfmt: true`

chancez replied:
This was promising, but I'm hitting permissions issues as root. I suspect it's something to do with apparmor, and that the binfmt-support package probably adds a profile to allow this. Any ideas how I could work around this?

chancez replied:
Ah nvm I see, I have to use `/proc/sys/fs/binfmt_misc/register`

AkihiroSuda commented on pkg/cidata/cidata.TEMPLATE.d/boot/05-rosetta-volume.sh:
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
```
Every shell script must have `set -eu`

AkihiroSuda commented on pkg/cidata/cidata.TEMPLATE.d/boot/05-rosetta-volume.sh:
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+if [ ! -d /tmp/rosetta-share ]; then
```
What's /tmp/rosetta-share ?

jandubois replied:
It is the the "Rosetta Directory Share" (see https://developer.apple.com/documentation/virtualization/running_intel_binaries_in_linux_vms_with_rosetta).

The sample code mounts it under `/tmp`, but I would prefer to mount it as `/mnt/rosetta` instead.

chancez replied:
Oops, I just forgot to update the path in the conditional to match the path where I mounted it.

cedws commented:
I don't have any comments on the changes other than nice work and LGTM, but do you have any suggested reading material on how this Rosetta acceleration works?

Thanks.

chancez commented on pkg/cidata/cidata.TEMPLATE.d/boot/05-rosetta-volume.sh:
```diff
@@ -0,0 +1,9 @@
+  # Cannot find a way to check that the virtiofs "device" is present, so just
+  # Try mounting it and hope for the best. If it fails, then it probably isn't
+  # a mac, or the virtio volume isn't being provided to the guest.
+  mount -t virtiofs vz-rosetta /mnt/lima-rosetta || true
```
Good thought, I'll see how difficult that is. I'm not sure if mounts works for anything besides bind-mounts, but if it does, then that would work.

balajiv113 replied:
Ah !! It won't work i think.
I missed a case, it will not work when we have mix of mountType. Using mountType: reversessh-fs and wanting to use rosetta.
This will not work as all the y.mounts are done based on mountType.

chancez replied:
We could add another mountType for this potentially. Not sure if it would be useful outside of this use-case yet though.

chancez replied:
I ended up continuing to use cidata, but we can revisit the approach later if needed.

chancez commented:
Okay, PR is updated based on feedback and I like it much better than the initial MVP. Let me know what you all think. I also rebased and moved it from draft->ready to review.

balajiv113 commented on pkg/vz/rosetta_directory_share.go:
```diff
@@ -0,0 +1,14 @@
+)
+
+func createRosettaDirectoryShareConfiguration(install bool) (*vz.VirtioFileSystemDeviceConfiguration, error) {
+	return nil, fmt.Errorf("Rosetta is unsupported on non-ARM64 hosts")
```
We should do a warning instead of error.

This is to make vz template runnable on both arm & x64

chancez replied:
Couldn't users just disable Rosetta on x86 in that case?

chancez replied:
I also think in that case, we probably can still return an error here, but we can turn it into a named error and check using `errors.Is` in the caller.

chancez replied:
Updated

balajiv113 commented on pkg/vz/vm_darwin.go:
```diff
@@ -344,6 +344,16 @@ func attachFolderMounts(driver *driver.BaseDriver, vmConfig *vz.VirtualMachineCo
+
+	if driver.Yaml.Rosetta.Enabled {
+		logrus.Info("Setting up Rosetta share")
+		directorySharingDeviceConfig, err := createRosettaDirectoryShareConfiguration(driver.Yaml.Rosetta.Install)
```
Just a thought,

Do we needed Rosetta.Install ? 
Can we directly install since Rosetta.Enabled is true already ?

chancez replied:
I'm not sure. I figured it would make sense to make it configurable in case users want to install Rosetta themselves. I already have it installed, so I'm not sure what it actually looks like if you try without it.

balajiv113 replied:
During installation nothing will pop-up or happen. It just like installation via command line. So i think it should be fine. 

But let's also hear others opinion on this.

AkihiroSuda replied:
> Can we directly install since Rosetta.Enabled is true already ?

:+1:

AkihiroSuda replied:
> Can we directly install since Rosetta.Enabled is true already ?

:+1: , assuming that `Enabled = true` doesn't show the macOS password popup as long as Rosetta is already installed on the macOS host.

balajiv113 replied:
@AkihiroSuda Your assumption is true. Only for installation the password pop-up will be shown. Not for every run of rosetta.

chancez replied:
PR updated

chancez commented:
@cedws 
> I don't have any comments on the changes other than nice work and LGTM, but do you have any suggested reading material on how this Rosetta acceleration works?
> 
> Thanks.

For how it works with Linux: https://developer.apple.com/documentation/virtualization/running_intel_binaries_in_linux_vms_with_rosetta

Rosetta in general: https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment

chancez commented:
There was also some discussion about whether or not this should be enabled by default in the other PR. What do people think?

chancez commented on pkg/limayaml/limayaml.go:
```diff
@@ -99,6 +107,10 @@ type NineP struct {
 }
 
+type VirtioFS struct {
+	Tag *string `yaml:"tag,omitempty" json:"tag,omitempty"`
```
Fixed

AkihiroSuda commented:
> There was also some discussion about whether or not this should be enabled by default in the other PR. What do people think?

The mount can be enabled by default, but binfmt should be probably opt-in for now.
The `vz.yaml` template can have `rosetta.binfmt: true` by default.

chancez commented:
> > There was also some discussion about whether or not this should be enabled by default in the other PR. What do people think?
> 
> The mount can be enabled by default, but binfmt should be probably opt-in for now. The `vz.yaml` template can have `rosetta.binfmt: true` by default.

Is that best done via `FillDefault` or should we change the field to `Disable` so the zero value is that Rosetta is enabled? (Disable: false)

AkihiroSuda commented:
> > > There was also some discussion about whether or not this should be enabled by default in the other PR. What do people think?
> > 
> > 
> > The mount can be enabled by default, but binfmt should be probably opt-in for now. The `vz.yaml` template can have `rosetta.binfmt: true` by default.
> 
> Is that best done via `FillDefault` or should we change the field to `Disable` so the zero value is that Rosetta is enabled? (Disable: false)

Lima itself should leave it disabled by default. The `vz.yaml` template file may have `rosetta.binfmt: true` line to override the built-in default value.

chancez commented:
@AkihiroSuda Sorry I meant for `rosetta.enabled`. 

Also what are your thoughts about the Rosetta installation behavior?

AkihiroSuda commented:
Looks like installing rosetta requires the root of the macOS host, so it should be disabled by default.

https://developer.apple.com/documentation/virtualization/running_intel_binaries_in_linux_vms_with_rosetta?language=objc
> Installing Rosetta is a one-time process per computer that requires the user to grant permission for the system to install Rosetta. If Rosetta is already installed in macOS, the system activates it for use under the Virtualization framework; if Rosetta isn’t installed, the framework downloads the installer from the network and performs the installation. The installer interactively prompts the user for authorization, and your app should handle any of the possible error conditions that could occur during the authorization, download, and installation process.

chancez commented:
@AkihiroSuda then it sounds like the PR as it is should be correct with the current defaults and `vz.yaml`. I've rebased.

chancez commented:
Getting a lint error on my `errRosettaUnsupported ` variable from golangci-lint, presumably because it's running without the build tags. I can add the needed build tags, but this probably means `golangci-lint` is not running on the new VZ code. We should update the GitHub action to specify the necessary build tags so it's getting properly linted.

AkihiroSuda reviewed (approved):
Thanks

balajiv113 commented on pkg/vz/rosetta_directory_share.go:
```diff
@@ -0,0 +1,12 @@
+//go:build darwin && !arm64
+// +build darwin,!arm64
```
I believe we need to use !no_vz buildtag in both rosetta files to avoid this issue 
https://github.com/lima-vm/lima/issues/1176

chancez replied:
Done

balajiv113 commented on pkg/vz/errors.go:
```diff
@@ -0,0 +1,8 @@
+//go:build darwin
+// +build darwin
```
Do we need this tags ?? 
errors.go can be generic irrespective of platform (We can move this https://github.com/lima-vm/lima/blob/master/pkg/vz/vz_driver_others.go#L13 as well here)

If we need this tag, then naming the file as errors_darwin.go should be sufficient and consistent across i believe

chancez replied:
Without build tags the variable is detected as unused in non-darwin builds.

chancez replied:
Renamed the file, and I updated build tags as more were required due to your previous comment: https://github.com/lima-vm/lima/pull/1155#discussion_r1029696381

AkihiroSuda commented on pkg/vz/vm_darwin.go:
```diff
@@ -345,6 +345,19 @@ func attachFolderMounts(driver *driver.BaseDriver, vmConfig *vz.VirtualMachineCo
+		if errors.Is(err, errRosettaUnsupported) {
+			logrus.Warnf("Unable to configure Rosetta: %s", err)
+		} else if err != nil {
+			return err
```
Probably, any error from `createRosettaDirectoryShareConfiguration` can be a warning.
e.g., failure of installation of Rosetta

chancez replied:
I'm still unsure of this. Is there other example cases where Lima fails to do setup and continues on anyways? It seems like continuing on would lead to unexpected results for users who may not notice the warnings, etc.

AkihiroSuda replied:
> Is there other example cases where Lima fails to do setup and continues on anyways? 

Yes, in a bunch of places (`git grep -F .Warn`)

e.g.,

https://github.com/lima-vm/lima/blob/32baa6216f89ce4aaacbc254d13f31535bffeb2d/pkg/start/start.go#L213

chancez replied:
Gotcha. I've updated it to warn instead.

AkihiroSuda reviewed (approved):
Thanks

AkihiroSuda commented:
Released beta.1 with Rosetta
https://github.com/lima-vm/lima/releases/tag/v0.14.0-beta.1
